src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/ClusterDetailsEventProcessor.java

```diff
@@ -125,11 +132,27 @@ public static NodeDetails getCurrentNodeDetails() {
 
   public static class NodeDetails {
 
-    private String id;
-    private String hostAddress;
-    private String role;
+    private final String id;
+    private final String hostAddress;
+    private final String role;
     private Boolean isMasterNode;
 
+    /**
+     * This constructor is provided expressly for testing purposes. In general, NodeDetails should only be
+     * retrieved through method calls such as {@link ClusterDetailsEventProcessor#getCurrentNodeDetails()}
+     * @param id The node's id
+     * @param hostAddress The host's address
+     * @param role The node's role e.g. data-node
+     * @param isMasterNode Whether or not the node is a master node
+     */
+    @VisibleForTesting
+    public NodeDetails(String id, String hostAddress, String role, Boolean isMasterNode) {
```

Comment:
We don't need to add a constructor specifically used by unit test. you can work around this by using a utility function API provided by another unit test. (createNodeDetailsMetrics)

Reply:
Great catch! I've improved the design as you suggested and will remember this pattern for the future.